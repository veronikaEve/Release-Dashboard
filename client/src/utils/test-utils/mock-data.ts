import { VersionData } from "../../@types";

export const mockVersionData: VersionData[] = [
  {
    associatedUser: "@johnSmith",
    date: "11/03/2022",
    versionFrom: "v7.8.2",
    versionTo: "v7.9.6",
    _id: "63122a332e9904df79c834ff",
  },
  {
    associatedUser: "@billy1234",
    date: "09/05/2022",
    versionFrom: "v7.7.10",
    versionTo: "v7.8.2",
    _id: "6313b230e6c802832ffeee01",
  },
];
