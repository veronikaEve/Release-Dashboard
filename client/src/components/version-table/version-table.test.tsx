import { render, screen, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import VersionTable from "./version-table";
import { mockVersionData } from "../../utils/test-utils/mock-data";

describe("VersionTable", () => {
  let props = {
    versionData: mockVersionData,
    setReleaseInfo: jest.fn,
  };

  test("should match snapshot", () => {
    const { asFragment } = render(<VersionTable {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  describe("table body", () => {
    test("row should have the 'selected' class if selected", async () => {
      const user = userEvent.setup();
      render(<VersionTable {...props} />);

      const table = screen.getByRole("table");
      const tableRow1 = within(table).getByTestId("table-row-1");

      expect(tableRow1).not.toHaveClass("selected");

      await user.click(tableRow1);

      expect(tableRow1).toHaveClass("selected");
    });

    test("should only allow one row to be selected", async () => {
      const user = userEvent.setup();
      render(<VersionTable {...props} />);

      const table = screen.getByRole("table");
      const tableRows = within(table).queryAllByRole("row");

      const tableRow0 = within(table).getByTestId("table-row-0");
      const tableRow1 = within(table).getByTestId("table-row-1");

      await user.click(tableRow0); // click once
      await user.click(tableRow1); // click again on a different element

      const selectedRows = tableRows.filter(
        (row) => row.className.includes("selected") // filter down rows to only those that have the "selected" className
      );

      expect(selectedRows).toHaveLength(1);
    });

    test("should not render rows if there is no versionData", () => {
      props.versionData = [];
      render(<VersionTable {...props} />);

      expect(screen.queryByTestId("version-table__body")).toBeNull();
    });
  });
});

// Do i need to mock and write tests for the api fetch?
