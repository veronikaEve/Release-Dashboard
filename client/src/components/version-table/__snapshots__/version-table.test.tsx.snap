// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VersionTable should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="table-wrapper"
  >
    <div
      class="table-wrapper__scroll"
    >
      <table
        class="version-table table table-sm table-bordered table-hover"
        data-testid="version-table"
      >
        <thead
          class="version-table__header"
        >
          <tr>
            <th>
              Date
            </th>
            <th>
              From
            </th>
            <th>
              To
            </th>
            <th>
              Excuted by
            </th>
          </tr>
        </thead>
        <tbody
          class="version-table__body"
          data-testid="version-table__body"
        >
          <tr
            class=""
            data-testid="table-row-0"
          >
            <td>
              11/03/2022
            </td>
            <td>
              v7.8.2
            </td>
            <td>
              v7.9.6
            </td>
            <td>
              @johnSmith
            </td>
          </tr>
          <tr
            class=""
            data-testid="table-row-1"
          >
            <td>
              09/05/2022
            </td>
            <td>
              v7.7.10
            </td>
            <td>
              v7.8.2
            </td>
            <td>
              @billy1234
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
